manager/state/raft/transport/peer.go

```diff
@@ -0,0 +1,153 @@
+
+	stopped chan struct{}
+	done    chan struct{}
+
+	mu   sync.Mutex
+	cc   *grpc.ClientConn
+	addr string
+
+	active       bool
+	becameActive time.Time
+}
+
+func newPeer(id uint64, addr string, tr *Transport) (*peer, error) {
+	cc, err := tr.dial(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create conn for %x with addr %s", id, addr)
+	}
+	return &peer{
+		id:           id,
+		addr:         addr,
+		cc:           cc,
+		tr:           tr,
+		msgc:         make(chan raftpb.Message, 4096),
+		stopped:      make(chan struct{}),
+		done:         make(chan struct{}),
+		active:       true,
+		becameActive: time.Now(),
+	}, nil
+}
+
+func (p *peer) send(ctx context.Context, m raftpb.Message) (err error) {
+	defer func() {
+		if err != nil {
+			p.tr.config.TransportRaft.ReportUnreachable(m.To)
+			p.mu.Lock()
+			p.active = false
+			p.mu.Unlock()
+		}
+	}()
+	select {
+	case p.msgc <- m:
+	case <-p.stopped:
+		return errors.New("peer stopped")
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		return errors.Errorf("peer is unreachable")
+	}
+	return nil
+}
+
+func (p *peer) update(conn *grpc.ClientConn, addr string) {
+	p.mu.Lock()
+	p.cc.Close()
+	p.cc = conn
+	p.addr = addr
+	p.mu.Unlock()
+}
+
+func (p *peer) conn() *grpc.ClientConn {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.cc
+}
+
+func (p *peer) address() string {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.addr
+}
+
+func (p *peer) resolveAddr(ctx context.Context, id uint64) (string, error) {
+	resp, err := api.NewRaftClient(p.conn()).ResolveAddress(ctx, &api.ResolveAddressRequest{RaftID: id})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to resolve address")
+	}
+	return resp.Addr, nil
+}
+
+func (p *peer) sendProcessMessage(ctx context.Context, m raftpb.Message) error {
+	_, err := api.NewRaftClient(p.conn()).ProcessRaftMessage(ctx, &api.ProcessRaftMessageRequest{Message: &m})
+	if err == nil {
+		if m.Type == raftpb.MsgSnap {
+			p.tr.config.TransportRaft.ReportSnapshot(m.To, raft.SnapshotFinish)
+		}
+		p.mu.Lock()
+		if !p.active {
+			p.active = true
+			p.becameActive = time.Now()
+		}
+		p.mu.Unlock()
+		return nil
+	}
+	p.mu.Lock()
+	p.active = false
+	p.mu.Unlock()
+	if m.Type == raftpb.MsgSnap {
+		p.tr.config.TransportRaft.ReportSnapshot(m.To, raft.SnapshotFailure)
+	}
+	p.tr.config.TransportRaft.ReportUnreachable(m.To)
+	if grpc.ErrorDesc(err) == membership.ErrMemberRemoved.Error() {
+		p.tr.config.TransportRaft.NodeRemoved()
+	}
+	return err
+}
+
+func (p *peer) processLoop(ctx context.Context) {
+	defer close(p.done)
+	for {
+		select {
+		case m := <-p.msgc:
+			ctx, cancel := context.WithTimeout(ctx, p.tr.config.SendTimeout)
+			err := p.sendProcessMessage(ctx, m)
+			cancel()
+			if err != nil {
+				log.G(ctx).WithError(err).Errorf("failed to send message %s to peer %x", m.Type, m.To)
+			}
+		case <-ctx.Done():
+			return
+		case <-p.stopped:
+			return
+		}
+	}
+}
+
+func (p *peer) stop() {
+	close(p.stopped)
+	<-p.done
```

Comment:
Seems like this could block for a long time if `p.msgc` has a lot of messages enqueued. Maybe `processLoop` should check if `p.stopped` is closed before selecting on `p.msgc`.